# Neuron DropOut Example

**In DeepLearning sometimes we tend to over apply the training in a way that we overfit the model. In this notebook I will try to prove that Neuron DropOut can make a model more accuarate in the end. The example will use tensorflow with keras and digits dataset.**

In [1]:
import tensorflow as tf
from tensorflow import keras
import pandas as pd

In [2]:
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

**First Model:**

In [3]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(100, activation='relu'),
    keras.layers.Dense(50, activation='relu'),
    keras.layers.Dense(25, activation='relu'),
    keras.layers.Dense(10, activation='sigmoid')
])

In [4]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [5]:
model.fit(x_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 3s 1ms/step - loss: 1.0194 - accuracy: 0.8166
Epoch 2/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2704 - accuracy: 0.9256
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1892 - accuracy: 0.9465
Epoch 4/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1557 - accuracy: 0.9547
Epoch 5/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1340 - accuracy: 0.9624
Epoch 6/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1205 - accuracy: 0.9656
Epoch 7/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1013 - accuracy: 0.9710
Epoch 8/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0955 - accuracy: 0.9721
Epoch 9/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0825 - accuracy: 0.9764
Epoch 10/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.075

In [6]:
from sklearn.metrics import confusion_matrix , classification_report
import numpy as np
y_pred = model.predict(x_test)
y_pred_classes = [np.argmax(element) for element in y_pred]

print("Classification Report: \n", classification_report(y_test, y_pred_classes))

313/313 [==============================] - 0s 675us/step
Classification Report: 
               precision    recall  f1-score   support

           0       0.31      0.99      0.47       980
           1       0.35      0.99      0.51      1135
           2       0.16      0.03      0.06      1032
           3       0.48      0.35      0.40      1010
           4       0.35      0.18      0.24       982
           5       0.96      0.95      0.95       892
           6       0.61      0.06      0.12       958
           7       0.53      0.02      0.04      1028
           8       0.98      0.78      0.87       974
           9       0.95      0.38      0.54      1009

    accuracy                           0.47     10000
   macro avg       0.57      0.47      0.42     10000
weighted avg       0.56      0.47      0.41     10000



**As we can see in the classification report our model was not so good as the epochs accuaracy parameter tended to show, their we got 97.8% accuracy, but with the tests our overall accuracy (f1-score) was only 47%.**
**With that we can conclude that First Model was not very good at all. Now let's try to use drop-out technique to improve our f1-score in a second model.**

**DropOut Model:**

In [7]:
modelDropOut = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(100, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(50, activation='relu'),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(25, activation='relu'),
    keras.layers.Dense(10, activation='sigmoid')
])

In [8]:
modelDropOut.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [9]:
modelDropOut.fit(x_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 3s 1ms/step - loss: 2.8803 - accuracy: 0.3189
Epoch 2/10
1875/1875 [==============================] - 2s 1ms/step - loss: 1.5217 - accuracy: 0.4512
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 1.2124 - accuracy: 0.5672
Epoch 4/10
1875/1875 [==============================] - 2s 1ms/step - loss: 1.0127 - accuracy: 0.6382
Epoch 5/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.9076 - accuracy: 0.6763
Epoch 6/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.8572 - accuracy: 0.6914
Epoch 7/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.8194 - accuracy: 0.7042
Epoch 8/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.7903 - accuracy: 0.7119
Epoch 9/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.7684 - accuracy: 0.7201
Epoch 10/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.743

In [10]:
y_pred = modelDropOut.predict(x_test)
y_pred_classes = [np.argmax(element) for element in y_pred]

print("Classification Report: \n", classification_report(y_test, y_pred_classes))

313/313 [==============================] - 0s 696us/step
Classification Report: 
               precision    recall  f1-score   support

           0       0.89      0.97      0.93       980
           1       0.99      0.98      0.98      1135
           2       0.50      0.88      0.64      1032
           3       0.93      0.88      0.91      1010
           4       0.94      0.85      0.89       982
           5       0.91      0.76      0.83       892
           6       0.94      0.95      0.94       958
           7       0.96      0.92      0.94      1028
           8       0.32      0.08      0.13       974
           9       0.78      0.93      0.85      1009

    accuracy                           0.83     10000
   macro avg       0.82      0.82      0.80     10000
weighted avg       0.82      0.83      0.81     10000



**Conclusion:**

**As we can see when doing the epochs we got the impression that our accuracy in 'modelDropOut' would be lesser then in 'model', but when we run our classification report we discover the contraty. The non-dropout model got an f1-score of 47% and the dropout model got an f1-score of 83%, way better then the first other model. With that we can check how the DropOut technique improves alot the accuracy in a model reducing it's overfitting.**